File: app/src/main/java/com/tribalfs/stargazers/ui/screens/main/MainActivity.java
package com.tribalfs.stargazers.ui.screens.main;

import android.content.Intent;
import android.os.Bundle;
import android.view.Window;

import androidx.activity.OnBackPressedCallback;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.splashscreen.SplashScreen;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentTransaction;
import androidx.recyclerview.widget.LinearLayoutManager;

import java.util.ArrayList;
import java.util.List;

import dev.oneuiproject.oneui.layout.DrawerLayout;
import dev.oneuiproject.oneui.layout.ToolbarLayout.Badge;
import dev.oneuiproject.oneui.widget.ScrollAwareFloatingActionButton;

import com.tribalfs.stargazers.R;
import com.tribalfs.stargazers.databinding.ActivityMainBinding;
import com.tribalfs.stargazers.ui.core.base.FragmentInfo;
import com.tribalfs.stargazers.ui.screens.customabout.CustomAboutActivity;
import com.tribalfs.stargazers.ui.screens.main.stargazerslist.StargazersListFragment;
import com.tribalfs.stargazers.ui.core.drawer.DrawerListAdapter;

public class MainActivity extends AppCompatActivity
        implements DrawerListAdapter.DrawerListener {
    private ActivityMainBinding mBinding;
    private FragmentManager mFragmentManager;
    private final List<Fragment> fragments = new ArrayList<>();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        SplashScreen.installSplashScreen(this);
        super.onCreate(savedInstanceState);
        setupSharedElementAnimation();
        mBinding = ActivityMainBinding.inflate(getLayoutInflater());
        setContentView(mBinding.getRoot());

        initFragmentList();
        initDrawer();
        initFragments();
        initOnBackPressed();
    }


    private void setupSharedElementAnimation() {
        Window window = getWindow();
        window.requestFeature(Window.FEATURE_ACTIVITY_TRANSITIONS);
    }

    private void initFragmentList() {
        fragments.add(new StargazersListFragment());
//        fragments.add(null);
//        fragments.add(new IconsFragment());
//        fragments.add(new AppPickerFragment());
    }

    private OnBackPressedCallback mBackPressedCallback = new OnBackPressedCallback(false) {
        @Override
        public void handleOnBackPressed() {
            onDrawerItemSelected(0);
            ((DrawerListAdapter)mBinding.drawerListView.getAdapter()).setSelectedItem(0);
        }
    };

    private void initOnBackPressed() {
        getOnBackPressedDispatcher().addCallback(this, mBackPressedCallback);
    }

    private void initDrawer() {
        mBinding.drawerLayout.setDrawerButtonIcon(getDrawable(dev.oneuiproject.oneui.R.drawable.ic_oui_info_outline));
        mBinding.drawerLayout.setDrawerButtonTooltip("App info");
        mBinding.drawerLayout.setDrawerButtonOnClickListener(v -> {
            startActivity( new Intent(MainActivity.this, CustomAboutActivity.class));
            mBinding.drawerLayout.setDrawerButtonBadge(new Badge.None());
        });

        mBinding.drawerListView.setLayoutManager(new LinearLayoutManager(this));
        mBinding.drawerListView.setAdapter(new DrawerListAdapter(this, fragments, this));
        mBinding.drawerListView.setItemAnimator(null);
        mBinding.drawerListView.setHasFixedSize(true);
        mBinding.drawerListView.seslSetLastRoundedCorner(false);
        mBinding.drawerLayout.setButtonBadges(new Badge.Dot(), new Badge.Dot());
        mBinding.drawerLayout.setDrawerStateListener((state) -> {
            if (state == DrawerLayout.DrawerState.OPEN) {
                mBinding.drawerLayout.setNavigationButtonBadge(new Badge.None());
            }
            return null;
        });
    }

    private void initFragments() {
        mFragmentManager = getSupportFragmentManager();
        FragmentTransaction transaction = mFragmentManager.beginTransaction();
        for (Fragment fragment : fragments) {
            if (fragment != null) transaction.add(R.id.main_content, fragment);
        }
        transaction.commit();
        mFragmentManager.executePendingTransactions();

        onDrawerItemSelected(0);
    }

    @Override
    public boolean onDrawerItemSelected(int position) {
        Fragment newFragment = fragments.get(position);
        FragmentTransaction transaction = mFragmentManager.beginTransaction();
        for (Fragment fragment : mFragmentManager.getFragments()) {
            transaction.hide(fragment);
        }
        transaction.show(newFragment).commit();

        if (newFragment instanceof FragmentInfo) {
            if (!((FragmentInfo) newFragment).isAppBarEnabled()) {
                mBinding.drawerLayout.setExpanded(false, false);
                mBinding.drawerLayout.setExpandable(false);
            } else {
                mBinding.drawerLayout.setExpandable(true);
                mBinding.drawerLayout.setExpanded(false, false);
            }
            mBinding.drawerLayout.setTitle(((FragmentInfo) newFragment).getTitle());
            if (newFragment instanceof StargazersListFragment) {
                mBinding.drawerLayout.setExpandedSubtitle("Pull down to refresh");
            }
        }
        mBinding.drawerLayout.setDrawerOpen(false, true);

        mBackPressedCallback.setEnabled(position != 0);

        return true;
    }

    public DrawerLayout getDrawerLayout() {
        return mBinding.drawerLayout;
    }

    public ScrollAwareFloatingActionButton getFab(){
        return mBinding.fab;
    }
}

File: app/src/main/java/com/tribalfs/stargazers/ui/screens/main/MainActivity.kt
package com.tribalfs.stargazers.ui.screens.main

import android.content.Intent
import android.os.Bundle
import android.util.LayoutDirection
import android.view.Menu
import android.view.MenuItem
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.res.ResourcesCompat
import androidx.core.splashscreen.SplashScreen.Companion.installSplashScreen
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.ViewModelProvider
import androidx.navigation.NavController
import androidx.navigation.fragment.NavHostFragment
import com.tribalfs.stargazers.R
import com.tribalfs.stargazers.data.StargazersRepo
import com.tribalfs.stargazers.databinding.ActivityMainBinding
import com.tribalfs.stargazers.ui.core.util.launchAndRepeatWithLifecycle
import com.tribalfs.stargazers.ui.screens.customabout.CustomAboutActivity
import com.tribalfs.stargazers.ui.screens.main.core.navigation.AppNavigation
import com.tribalfs.stargazers.ui.screens.main.core.navigation.MainNavigationDelegate
import com.tribalfs.stargazers.ui.screens.settings.SettingsActivity
import dev.oneuiproject.oneui.ktx.setBadge
import dev.oneuiproject.oneui.layout.Badge
import dev.oneuiproject.oneui.layout.NavDrawerLayout
import dev.oneuiproject.oneui.utils.ActivityUtils
import kotlinx.coroutines.flow.collectLatest

//TODO/s:
// 1. navigate using navigation lib ✔
// 2. Add fragments for each repo ✔
// 3. Allow adding repo in drawer
class MainActivity : AppCompatActivity(),
    AppNavigation by MainNavigationDelegate() {

    private lateinit var mBinding: ActivityMainBinding
    private lateinit var navController: NavController

    private lateinit var mainViewModel: MainViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        installSplashScreen()
        super.onCreate(savedInstanceState)
        mBinding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(mBinding.root)
        mBinding.drawerLayout.setTitle(getString(R.string.app_name))
        setupNavigation()

        initViewModel()

        launchAndRepeatWithLifecycle(Lifecycle.State.RESUMED){
            mainViewModel.updateAvailableStateFlow.collectLatest {
                mBinding.drawerLayout.setHeaderButtonBadge(if (it) Badge.DOT else Badge.NONE)
            }
        }

        if (savedInstanceState == null) {
            mainViewModel.checkUpdate()
        }
    }

    private fun setupNavigation() {
        mBinding.drawerLayout.apply {
            setHeaderButtonIcon(
                ResourcesCompat.getDrawable(resources,
                dev.oneuiproject.oneui.R.drawable.ic_oui_settings_outline, theme)
            )
            setHeaderButtonTooltip("Stargazers Settings")
            setHeaderButtonOnClickListener {
                ActivityUtils.startPopOverActivity(
                    this@MainActivity,
                    Intent(this@MainActivity, SettingsActivity::class.java),
                    null,
                    ActivityUtils.POP_OVER_POSITION_TOP or
                            (if (isRTL) ActivityUtils.POP_OVER_POSITION_RIGHT else ActivityUtils.POP_OVER_POSITION_LEFT)
                )
            }
            setNavRailContentMinSideMargin(14)
        }

        val navHostFragment = supportFragmentManager.findFragmentById(R.id.nav_host_main) as NavHostFragment
        navController = navHostFragment.navController

        initNavigation(mBinding.drawerLayout, mBinding.drawerListView, navController)
    }

    private fun initViewModel() {
        val stargazersRepo = StargazersRepo.getInstance(this)
        val viewModelFactory = MainViewModelFactory(stargazersRepo)
        mainViewModel = ViewModelProvider(this, viewModelFactory)[MainViewModel::class.java]
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.menu_main_activity, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.menu_app_info) {
            startActivity(Intent(this@MainActivity, CustomAboutActivity::class.java))
            return true
        }
        return false
    }

    private val isRTL: Boolean get() = resources.configuration.layoutDirection == LayoutDirection.RTL

    val drawerLayout: NavDrawerLayout get() = mBinding.drawerLayout

    companion object{
        const val KEY_REPO_NAME = "repoName"
    }
}
